collector/ccr_stats.go

```diff
@@ -0,0 +1,448 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/go-kit/kit/log/level"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"net/url"
+	"path"
+	"strconv"
+)
+
+// CCRStats information struct
+type CCRStats struct {
+	logger log.Logger
+	client *http.Client
+	url    *url.URL
+
+	up                              prometheus.Gauge
+	totalScrapes, jsonParseFailures prometheus.Counter
+	ccrMetrics                      []*IndexCCRStatsShardMetric
+}
+
+type IndexCCRStatsShardMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(shardCCRStats IndexCCRStatsShardResponse) float64
+	Labels func(indexCCRStatsShard IndexCCRStatsShardResponse) []string
+}
+
+var (
+	defaultCCRStatsLabels = []string{"follower_index", "leader_index", "remote_cluster", "shard_id"}
+	defaultCCRLabelValues = func(indexCCRStatsShard IndexCCRStatsShardResponse) []string {
+		return []string{indexCCRStatsShard.FollowerIndex, indexCCRStatsShard.LeaderIndex, indexCCRStatsShard.RemoteCluster, strconv.Itoa(int(indexCCRStatsShard.ShardID))}
+	}
+)
+
+// NewCCRStats defines CCRStats Prometheus metrics
+func NewCCRStats(logger log.Logger, client *http.Client, url *url.URL) *CCRStats {
+	return &CCRStats{
+		logger: logger,
+		client: client,
+		url:    url,
+
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: prometheus.BuildFQName(namespace, "ccr_stats", "up"),
+			Help: "Was the last scrape of the ElasticSearch ccr stats endpoint successful.",
+		}),
+		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, "ccr_stats", "total_scrapes"),
+			Help: "Current total ElasticSearch ccr stats scrapes.",
+		}),
+		jsonParseFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, "ccr_stats", "json_parse_failures"),
+			Help: "Number of errors while parsing JSON.",
+		}),
+		ccrMetrics: []*IndexCCRStatsShardMetric{
```

Comment:
Metrics are now declared at the package level. This allows us to easily find any that aren't being used.